### PR TITLE
OPS-5046 support spark 3.4.0 on SPOK

### PR DESCRIPTION
**Pull request by xuzeng012**
spark 3.4.0 made a change to spark type `CharType` which requires the constructor to take a parameter `length` so we are using spark's internal method to take care of all that parsing

### CODE REVIEW

File: sparkly/utils.py [hunk 1]
```diff
@@ -176,73 +176,14 @@ def parse_schema(schema):
         sparkly.exceptions.UnsupportedDataType: Cannot parse type from string: "unsupported"
     """
     field_type, args_string = re.match('(\w+)<?(.*)>?$', schema).groups()
-    args = _parse_args(args_string) if args_string else []
 
-    if field_type in ATOMIC_TYPES:
-        return ATOMIC_TYPES[field_type]()
-    elif field_type in COMPLEX_TYPES:
-        return COMPLEX_TYPES[field_type](*args)
+    if field_type in T._all_atomic_types or field_type in T._all_complex_types:
```
Comment (srstrickland) [inline sparkly/utils.py:180]: why `re.match` at all, and why check the top-level `field_type`?  why not just defer entirely?

Comment (xuzeng012) [reply sparkly/utils.py:180]: ah you are right, so just use schema as is?

File: sparkly/utils.py [hunk 2]
```diff
@@ -175,73 +175,7 @@ def parse_schema(schema):
...
-    return T.ArrayType(parse_schema(args[0]))
-
-
-COMPLEX_TYPES = {
-    'map': _init_map,
-    'struct': _init_struct,
-    'array': _init_array,
-}
-
+    return T._parse_datatype_string(schema)
```
Comment (srstrickland) [inline sparkly/utils.py:178]: It's possible that we may want to catch any errors, and rethrow them as our custom `UnsupportedDataType` exception.

**Review by srstrickland — COMMENTED** [review]
tests are broken, it seems because the `Dockerfile` is not properly set up to install jdk8.  this needs to be fixed.  I recommend testing locally using `make test`, or for faster iteration, `make dev` then run tests with `tox` when you're inside the container.

File: sparkly/utils.py (hunk 2 above)
Comment (xuzeng012) [reply sparkly/utils.py:178]: you mean something like this 
```
    try:
        return T._parse_datatype_string(schema)
    except:
        message = 'Cannot parse type from string: "{}"'.format(field_type)
        raise UnsupportedDataType(message)
```

**Comment by srstrickland** [comment]
Looks good, but let's make sure!  We can add a `spark34` section to [tox.ini](https://github.com/tubular/sparkly/blob/master/tox.ini) to force our test suite to run on spark 3.4.  just copy the `spark33` section and modify the `pyspark` version.

**Comment by srstrickland** [comment]
Looks like there are several failures with `3.4.0`:

```
tests/integration/test_catalog.py ..............F...............F.       [ 19%]
tests/integration/test_functions.py .....................                [ 31%]
tests/integration/test_instant_testing.py ..                             [ 32%]
tests/integration/test_reader.py ......                                  [ 36%]
tests/integration/test_session.py ....                                   [ 38%]
tests/integration/test_testing.py .........                              [ 44%]
tests/integration/test_writer.py .....FF.....................            [ 61%]
tests/unit/test_instant_testing.py ....                                  [ 63%]
tests/unit/test_reader.py ...........                                    [ 70%]
tests/unit/test_session.py ...............                               [ 79%]
tests/unit/test_testing.py ..........                                    [ 85%]
tests/unit/test_utils.py .................                               [ 95%]
tests/unit/test_writer.py ........                                       [100%]
```

it would seem that more work is required here to make everything work.

File: tests/integration/base.py [hunk 3]
```diff
@@ -26,7 +26,7 @@ class SparklyTestSession(SparklySession):
     packages = [
         'com.datastax.spark:spark-cassandra-connector_2.12:3.2.0',
         'org.elasticsearch:elasticsearch-spark-30_2.12:7.17.8',
-        'org.apache.spark:spark-sql-kafka-0-10_2.12:3.3.1',
+        'org.apache.spark:spark-sql-kafka-0-10_2.12:{}'.format(os.getenv('spark_version')),
```
Comment (srstrickland) [inline tests/integration/base.py:29]: if you `import pyspark`, the version is available via `pyspark.__version__`.

File: tox.ini [hunk 4]
```diff
@@ -24,6 +24,8 @@ deps =
     -rrequirements_dev.txt
     -rrequirements_extras.txt
     pyspark==3.2.3
+setenv =
+    spark_version = 3.2.3
```
Comment (srstrickland) [inline tox.ini:28]: clever workaround, but let's drop this in favor of using `pyspark.__version__`

**Review by srstrickland — APPROVED** [review]

**Review by srstrickland — APPROVED** [review]

**Pull request merged**